### PR TITLE
Cross model relations: Offers

### PR DESCRIPTION
**Pull request by SimonRichardson**
Adding cross model offer add offer, remove offer and list offer.
The following also includes integration test around add and
removing offers.

## QA

You need to prepare a juju before doing the following:

```console
juju bootstrap lxd test --no-gui
```

Ensure that the example works by doing:

```console
tox -e example -- examples/crossmodel.py
```

Ensure that integration also works:

```console
tox -e integration -- tests/integration/test_crossmodel.py
```

### CODE REVIEW

**Comment by SimonRichardson** [comment]
https://trello.com/c/i2fhLvFa/1621-cmr-pylibjuju-offer-application

**Comment by mitechie** [comment]
Thanks for the updates. This looks solid and works. I'd like us to look into pulling some common parsing things out into a shared space just so we're not maintaining the regex for an application over and over and the like. Some comments mostly on just structure/organization. I think I struggled a bit with the number of things, named with very much similar names, being used in different ways. I've got some suggestions that are really just clarification-type stuff.

File: juju/model.py [hunk 1]
```diff
@@ -1881,6 +1881,35 @@ def charmstore(self):
...
+        Offer a deployed application using a series of endpoints for use by
+        consumers.
+
+        @param endpoint: holds the application and endpoint you want to offer
+        @param offer_name: over ride the offer name to help the consumer
+        """
+        controller = await self.get_controller()
+        return await controller.offer(self.info.uuid, endpoint, offer_name)
+
+    async def offers(self):
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply juju/model.py]: Yeah, I can do that!

File: tests/unit/test_offerendpoint.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        for name, case in cases.items():
+            self.assertEqual(parse_offer_url(name), case)
+
+    def test_parse_offer_url_failures(self):
+        cases = {"controller:modelname": "application offer URL is missing application",
+                 "controller:user/modelname": "application offer URL is missing application",
+                 "modelname": "application offer URL is missing application",
+                 "/user/modelname": "application offer URL is missing application",
+                 "modelname.applicationname@bad": "application name applicationname@bad not valid",
+                 "user[bad/model.application": "user name user[bad not valid",
```
Comment (mitechie) [inline tests/unit/test_offerendpoint.py]: are these brackets right? user[bad ? with no close?

Comment (SimonRichardson) [reply tests/unit/test_offerendpoint.py]: Yeah, it just an invalid character in the username

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**